Add keybind for cycling grid type

OliBomby opened the pull request:
Requires:
- [x] https://github.com/ppy/osu/pull/26309
- [x] https://github.com/ppy/osu/pull/26310
- [x] https://github.com/ppy/osu/pull/26311
- [x] https://github.com/ppy/osu/pull/26313

Using the tool boxes on the right frequently is quite annoying, so I also added a keybind for cycling the grid type (H). I chose this key because its right next to the grid spacing cycle hotkey (G) and unused. It makes sense to group the grid property changing hotkeys physically close on the keyboard. An alternative hotkey could be like Shift+G or Shift+G, Ctrl+G is taken.

There's also a test for the grid placement tool in here which apparently got left behind when I was splitting up this PR into parts. I guess it can't hurt to have it so here it is.

https://github.com/ppy/osu/assets/17460441/cf63348d-0fdc-4170-b497-bae3c69cee7a

OliBomby committed:
use G to change grid type

OliBomby committed:
fix grid test

OliBomby committed:
Add TestGridFromPoints

OliBomby committed:
Add back the old keybind for cycling grid spacing

bdach commented:
1.3k line diffstat moves this even lower on review priority list than it already was.

Did everything here *need* to be in a single pull?

OliBomby commented:
Right I forgot to split it up to make review more palatable.

OliBomby committed:
Fix grid type cycling not taking into account the radio button selection

OliBomby committed:
Update grid placement tool test

I somehow missed this test when splitting up PRs so here it is now

OliBomby committed:
Rename "Cycle grid display mode" to "Cycle grid spacing"

The "display mode" is easy to confuse with grid type, so I renamed it to literally the grid property it affects.

bdach commented:
What's with the history here?

Please force-push that away because whatever that diffstat is, it's not worth 50 commits. If you chain PRs correctly the commits from previous PRs in the chain should go away when merging master into PRs later down the chain; and here it appears they didn't. So I'd rather this was fixed up, even if by squashing, because I don't want to see all that stuff in blame.

OliBomby commented:
I'm struggling with git here to fix this history

peppy commented:
I dunno what's going on with the new commits but this should be fixed by squashing down to a single commit.

OliBomby commented:
Ok its fixed

peppy commented:
I guess cycling is okay and saves on used hotkeys, but I'd bet the first thing we get asked for next is per-grid hotkeys rather than a cycle. Any thoughts on this?

peppy committed:
Change shortcut to `Shift`+`G`

peppy committed:
Simplify cycle logic

peppy committed:
Fix bindings being clobbered

peppy commented:
I've pushed some changes and fixes. @OliBomby please confirm you're okay with them.

Importantly, you have to make sure to add new bindings to the end of the enum, else it offsets game bindings game-wide and breaks everything (see https://github.com/ppy/osu/pull/26303/commits/b62d9f8696206b9779c7d7286d974cd4e41fe581).

peppy committed:
Update test

OliBomby committed:
Explicitly assert specific grid type

OliBomby commented on osu.Game.Rulesets.Osu.Tests/Editor/TestSceneOsuEditorGrids.cs:
```diff
@@ -189,5 +191,92 @@ private void nextGridSizeIs(int size)
+            AddUntilStep("rectangular grid visible", () => this.ChildrenOfType<RectangularPositionSnapGrid>().Any());
+            gridActive<RectangularPositionSnapGrid>(true);
+
+            nextGridTypeIs<TriangularPositionSnapGrid>();
```
I've added an assert to explicitly assert the specific grid type. Now it should fail on the assert if it has the wrong grid type.

bdach committed:
Merge branch 'master' into grids

bdach committed:
Remove unused field

Not sure why inspectcode is quiet about this?
